### ISSUE TITLE
yum update not working if lenght of package name + package version is too long

### ISSUE THREAD

**Issue by Spredzy** [ansible/ansible-modules-core#2802]
##### ISSUE TYPE

Bug Report
##### COMPONENT NAME

yum module
##### ANSIBLE VERSION

1.9.4
##### SUMMARY

If the the combo of package name + package version is too long, yum update (`yum: name=toto state=latest update_cache=yes`) will not work.

Example : 

```
dbus-libs.x86_64                     1:1.6.12-13.el7                   base
PACKAGE.x86_64                        0.0.201601130408git9dd37204-1.el7.centos
                                                                       REPO-devel
device-mapper.x86_64                 7:1.02.107-5.el7                  base
```

This is due to the fact that in order to decide for ansible to do an update or not it will : 
1. Run `yum -d 2 -y check-update`
2. Iterate over each line of the output
3. Do a `line = line.split()`
4. Check if the `len(line) != 3`, if so do nothing.

The issue is that the output of the line is too long and hence understood at two lines.

Line 1 : 'PACKAGE.x86_64                        0.0.201601130408git9dd37204-1.el7.centos'
Line 2 : '                                                                      REPO-devel'

So it will always do nothing as `line.split()` is never equal to 3.

Code responsible for this is [1]: 

``` python
# ignore irrelevant lines
# FIXME... revisit for something less kludgy
if '*' in line or len(line) != 3 or '.' not in line[0]:
    continue
```

[1] https://github.com/ansible/ansible-modules-core/blob/devel/packaging/os/yum.py#L777-L780

**Comment by Spredzy** [comment]
The first condition `'*' in line` could be removed by using `yum -q`, either globally either using `rc, out, err = module.run_command(yum_basecmd + ['-q', 'check-update'])`

The second conditon `len(line) != 3`, I think that only improving the parsing could work. I have no suggestion atm.

For my case using that did work. Will try with longer strings to find out the limit.

``` python
if not update_all:
      rc, out, err = module.run_command(yum_basecmd + ['-q', 'check-update'] + items)
else:
      rc, out, err = module.run_command(yum_basecmd + ['-q', 'check-update'])
```

The third condition `'.' not in line[0]` I am not sure what we want to verify with that. If any one know the use case, feeback will be welcome.

As I am not familiar enough with Ansible code, my suggestion might or might not apply, if not let me know.

**Comment by mscherer** [comment]
On what version do you see the problem ?

**Comment by Spredzy** [comment]
Sorry I missed to specify that.

```
#> ansible --version
ansible 1.9.4
#> cat /etc/fedora-release
Fedora release 23 (Twenty Three)
```

The node I am provisioning is running CentOS 7

**Comment by NathalieCBC** [comment]
We're facing this too. Sadly this seems to be a "feature" of yum, which they are not about to modify anytime soon (as ref in https://bugzilla.redhat.com/show_bug.cgi?id=584525). Yum will split lines at package name level or version level. Since indeed the output is split as:
available_updates = out.split('\n')
this will fail for some packages with long names or long versions names....
We've managed to make it work using this from line 766:

``` pyton
        available_updates = out.split('\n')
        # build update dictionary
        available_updates_length = len(available_updates)
        line_index = 0
        while line_index < available_updates_length:
            line = available_updates[line_index].split()
            # ignore irrelevant lines
            # FIXME... revisit for something less kludgy
            if len(line) != 3 and (line_index+1 < (available_updates_length -1)):
                if (len(line) + len(available_updates[line_index+1].split())) == 3:
                    line.extend(available_updates[line_index+1].split())
                    line_index += 1

            line_index += 1
            if '*' in line or len(line) != 3 or '.' not in line[0]:
                continue
            else:
                pkg, version, repo = line
                name, dist = pkg.rsplit('.', 1)
                updates.update({name: {'version': version, 'dist': dist, 'repo': repo}})
```

I can't say it is any less of a kludge... essentially, it is counting how long the line is and if shorter than one, it adds up the next line. It feels like this is something that should be fixed by a better output from yum but in the meanwhile, maybe this can help. There might also be more corner cases of line splitting not covered, this hasn't been thoroughly tested but it worked for us.

This is in ansible 2.0.1.0.

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/jctanner/ansibullbot/blob/ISSUE_TRIAGE_CLASS/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
This repository has been locked. All new issues and pull requests should be filed in https://github.com/ansible/ansible

Please read through the <a href="http://docs.ansible.com/ansible/dev_guide/repomerge.html#move-issues-and-prs-to-new-repo">repomerge</a> page in the dev guide. The guide contains links to tools which automatically move your issue or pull request to the ansible/ansible repo.

**Comment by bcoca** [comment]
closing in favor of https://github.com/ansible/ansible/issues/20043 which is in the new unified repo
